### DISCUSSION TITLE
Yii3: Add a common failure handling pattern to the middlewares.

### DISCUSSION THREAD

**Discussion by olegbaturin** [yiisoft/docs#213] [Ideas]
Sometimes middlewares stop processing request with their own error and return themselves a simple error response with only status and status text as the body. But error response is part of the application, and middleware must provide a way to customize it. So, the proposal is to make a common pattern for handling middleware errors from the app.

Below is a list of packages and classes that currently return error responses. Some of them have the failure handler, but others do not.
```
# auth
Authentication
# csrf
CsrfMiddleware
# rate-limiter
LimitRequestsMiddleware
# request-body-parser
RequestBodyParser
# router
Router
# yii-http
Application
# yii-middleware
IpFilter

```

Examples of failure handlers implemented well
```
# request-body-parser
RequestBodyParser - the failure handler has a context and implements its own interface
# auth
Authentication - the middleware has a default failure handler class to create a response
# csrf
CsrfMiddleware - the middleware creates an error response itself
```

To ensure uniform configuration, the constructor parameter should be named the same in all middlewares, **failureHandler** of **fallbackHandler**.

Example of configuring failure handlers in the middleware constructor
```php
CsrfMiddleware::class => [
    'class' => CsrfMiddleware::class,
    '__construct()' => [
        'failureHandler' => Reference::to(CsrfFailureHandler::class),
    ],
],

Authentication::class => [
    'class' => Authentication::class,
    '__construct()' => [
        'failureHandler' => Reference::to(AuthRequestFailureHandler::class),
    ],
],

```

**Comment by xepozz** [comment]
We can also go further and make it as a contract for all frameworks' middlewares
